submissions/hisbvdis/oop-exercise/index.js

```diff
@@ -0,0 +1,99 @@
+/* Refer to https://github.com/OleksiyRudenko/a-tiny-JS-world for the task details
+   Complete the below for code reviewers' convenience:
+
+   Code repository: https://github.com/hisbvdis/a-tiny-JS-world
+   Web app: _put project's github pages URL here_
+   */
+
+// ======== OBJECTS DEFINITIONS ========
+// Define your objects here
+class Creature {
+  constructor(species, name, gender, saying, legs) {
+    this.species = species;
+    this.name = name;
+    this.gender = gender;
+    this.legs = legs;
+    this.saying = saying;
+    this.props = [
+      { prop: "species", preface: "I am a" },
+      { prop: "name", preface: "My name is" },
+      { prop: "gender", preface: "I am a" },
+      { prop: "saying", preface: "I can say —" },
+      { prop: "legs", preface: "I have legs:" },
+    ];
+  }
+
+  getProps() {
+    return this.props
+      .map(({ prop, preface }) => `${preface} ${this[prop]}`)
+      .join("; ");
+  }
+}
+
+class Human extends Creature {
+  constructor(name, gender, saying) {
+    super("human", name, gender, saying, 2);
+    this.hands = 2;
+    this.props = this.props.concat({prop: "hands", preface: "I have hands:"});
```

Comment:
(1) ES6 offers a different pattern for object literal extension:

`obj = {...obj, ...extraProps, ...{yetAnotherProp1: 1, yetAnotherProp2: 2}}`

(2) The matter is that class should handle on its own the properties it owns immediately.
There should be good reasons to delegate handling to the base class. There are no such reasons in this task.
Use method overloading and explicit call to an inherited version of the method via extended `super` API.
Try to keep the elegancy of property-name/preface approach.

Reply:
> (1) ES6 offers a different pattern for object literal extension:
> 
> `obj = {...obj, ...extraProps, ...{yetAnotherProp1: 1, yetAnotherProp2: 2}}`

Коммит: 7f1b42f

Reply:
> Use method overloading and explicit call to an inherited version of the method via extended `super` API

Коммит:  5aa89e8

Reply:
You need to spread an object even if it consists of a single entry (key+prop) for this to work. Check the examples above.
![image](https://user-images.githubusercontent.com/8686631/191362819-e8e8c8be-58d8-4434-9e25-6eb89d33dd38.png)

Reply:
> > (2)...
> > Use method overloading and explicit call to an inherited version of the method via extended `super` API
> 
> Не очень хорошо понял этот момент. Постарался сделать так, как понял. Проверил через `console.log()` — при создании экземпляров вроде вызывается дочерний метод.
> 
> Коммит: [5aa89e8](https://github.com/kottans/frontend-2022-homeworks/commit/5aa89e8116eebbe4d3139ddea69b85d7909e73ab)

Yes, and it can call inherited version via `super.methodName` and add anything on top of that to build required outcome.

Reply:
@OleksiyRudenko 

> You need to spread an object even if it consists of a single entry (key+prop) for this to work. Check the examples above. ![image](https://user-images.githubusercontent.com/8686631/191362819-e8e8c8be-58d8-4434-9e25-6eb89d33dd38.png)

В примере выше Вы указали распространение объекта
`obj = {...obj, ...extraProps, ...{yetAnotherProp1: 1, yetAnotherProp2: 2}}`

Но ведь у меня "this.props" — это массив.  Массив объектов. И мне нужно добавить в массив ещё один объект. 

К сожалению, не понимаю, как сдесь можно использовать распространение объекта.  Ведь объекты в массиве хранятся в виде объектов

P.S. Может Вы подумали, что у меня `this.props` — это объект?

Reply:
> > > (2)...
> > > Use method overloading and explicit call to an inherited version of the method via extended `super` API
> > 
> > 
> > Не очень хорошо понял этот момент. Постарался сделать так, как понял. Проверил через `console.log()` — при создании экземпляров вроде вызывается дочерний метод.
> > Коммит: [5aa89e8](https://github.com/kottans/frontend-2022-homeworks/commit/5aa89e8116eebbe4d3139ddea69b85d7909e73ab)
> 
> Yes, and it can call inherited version via `super.methodName` and add anything on top of that to build required outcome.

@OleksiyRudenko , 

Из фразы `and add anything on top of that to build required outcome` я понял, что метод дочернего класса должен как-то изменить (дополнить) `this.props` родительского класса и вывести их.

К сожалению, приходится гадать.

Может изменять `this.props` нужно в дочернем методе?

Коммит: ca97d8d

Reply:
> P.S. Может Вы подумали, что у меня this.props — это объект?
Yes, that was my mistake. Your code is correct in these regards.

Reply:
> Из фразы and add anything on top of that to build required outcome я понял, что метод дочернего класса должен как-то изменить (дополнить) this.props родительского класса и вывести их.

Rather add. Changing properties that a class doesn't own immediately is in general a bad practice.
In this specific case changed `props` property will not be used by the child class. Instead the base class will handle child's property (to create an output string) it knows nothing and shouldn't know nothing about, implicitly extending base class functionality. Extension should only work one way: parent to child in all regards.

Reply:
Imagine we want to change how `hands` are reprsented internally (e.g. we want to store an object denoting count of left and right hands instead of simple number), still output would be a simple count of all hands. If we do this change we then are forced to change the base class (conversion of `hands` property into a number of hands). This breaks encapsulation and OOP principles. Besides we may not be allowed to change the base class at all.
So, a that owns a property should handle it on its own, end to end. Child classes may borrow (via inheritance or explicit calls of inherited versions) anything from base classes chain, but never from child to any of parents in the inheritance chain.

Reply:
@OleksiyRudenko ,

Может так?
Я Вызываю родительский метод и дополняю его данными из дочернего класса.

P.S. Мне очень жаль, что я не могу понять задачу. Видимо, не хватает понимания ООП.

Коммит:  ce36f35